### ISSUE TITLE
Reflection in CUE API

### ISSUE THREAD

**Issue by burdiyan** [cue-lang/cue#3378]
**Is your feature request related to a problem? Please describe.**

**TL;DR:** Have a way to [walk schemas](https://cuelang.org/docs/howto/walk-schemas-using-go-api/) from within CUE itself, without having to use the Go API.

---

I'm currently comparing CUE with [Apple's pkl language](https://pkl-lang.org) for the purposes of language-agnostic schema definitions and verifications. As far as I understand pkl, unlike CUE, lets you work with its own type definitions inside of its own code (i.e. reflection). E.g. this is how pkl implements code generation for its output languages: there's pkl code that reads types from other pkl code and uses string templating to generate the output code for the target programming language.

As far as I understand this is not possible with CUE. There's no way to write a CUE script/cmd-tool to read types from other CUE files and manipulate types as values. I guess it's possible to do using the Go API, but it's not possible from within CUE itself.

E.g. I'd like to define some schemas in CUE, write some code generation logic in CUE, and then use the `cue` command to generate the code, even without having Go toolchain installed, and without creating my own tools using Go to be distributed as compiled binaries.

**Describe the solution you'd like**
Expose some reflection APIs to CUE itself.

**Describe alternatives you've considered**
Write some custom tool using Go's CUE API to read the CUE schema and output some code.

**Comment by myitcv** [comment]
@burdiyan thanks for the question and welcome! 

Please can you share a pseudo-code like example of the kind of thing you'd be looking to do using such a feature?

**Comment by burdiyan** [comment]
Roughly speaking something like that:

```cue
package schema

import "text/template"
import "reflect"

#Person: {
	firstName: string
	lastName:  string
	age:       int
}

tpl: template.Execute("The type of Person.firstName is {{ .Fields.firstName.Type }}", reflect.TypeOf(#Person))
```

And then I'd `cue eval schema.cue --out text -e tpl` (or I'd write some `_tool.cue` file to make it a dedicated command), and I'd expect to see "The type of Person.firstName is string". Instead of this text I'd actually have some useful template to generate some code, let's say in TypeScript, of whatever.

The thing is, I want to be able to manipulate CUE schemas without having to use Go API.

It would also be great to extract doc comments from CUE types, so I could also use the comments to generate some docs or something like that.

Of course all of this could be done from within the Go API, but doing it from CUE itself has a nice "scripting" feel to it: no need for extra toolchains, no need to compile anything, no binaries to deal with.

**Comment by mvdan** [comment]
Note that we are adding some reflect builtins via this proposal, such as `exists` and `allows`: https://github.com/cue-lang/cue/discussions/4386

@burdiyan do you still see a gap if these were to be added?

**Comment by mvdan** [comment]
I'm also not sure I follow your earlier example with `reflect.TypeOf`. That's quite a large API surface on the Go side, so porting this to the CUE standard library does not make sense to me. It would be best to focus on clearer feature requests, such as being able to iterate over optional or required fields in a struct in CUE.

**Comment by mvdan** [comment]
Closing for now per the above, but happy to reopen if you can share new information. Thanks!

**Comment by burdiyan** [comment]
@mvdan my idea with this proposal was to be able to implement code generation within CUE itself, without having to use the Go API. For that two things are needed: a templating mechanism with replacements, and a way to introspect the type information. CUE already has the first part, but it doesn't have type introspection.

E.g. having a CUE definition like:

```cue
#Person: {
  name: string
}
```

there's no way within CUE itself to know fields of a Person nor the types of those fields.

I guess that's why projects like [Hof](https://github.com/hofstadter-io/hof) are using CUE as a DSL, instead of taking advantage of its own type-system.

**Comment by mvdan** [comment]
Understood. We would need a clearer proposal on how type introspection could actually work with a brief design and examples; I don't think a "TypeOf" example is clear enough as there's lots of open questions, such as what it would return or what its cost would be on large schemas/values.
